cpp/src/arrow/builder.cc

```diff
@@ -114,18 +114,51 @@ void ArrayBuilder::UnsafeAppendToBitmap(const uint8_t* valid_bytes, int32_t leng
     UnsafeSetNotNull(length);
     return;
   }
+
+  int byte_offset = length_ / 8;
+  int bit_offset = length_ % 8;
+  uint8_t bitset = null_bitmap_data_[byte_offset];
+
   for (int32_t i = 0; i < length; ++i) {
-    // TODO(emkornfield) Optimize for large values of length?
-    UnsafeAppendToBitmap(valid_bytes[i] > 0);
+    if (valid_bytes[i]) {
+      bitset |= (1 << bit_offset);
+    } else {
+      bitset &= ~(1 << bit_offset);
+      ++null_count_;
+    }
```

Comment:
See also https://github.com/apache/arrow/blob/5439b71586f4b0f9a36544b9e2417ee6ad7b48e8/cpp/src/arrow/util/bit-util.h#L67; you could avoid branching here if you wanted